### ISSUE TITLE
Automatic Generation of Identifier Based on Title

### ISSUE THREAD

**Issue by aubertc** [chdemko/pandoc-numbering#15]
TLDR: I suggest making the auto-identifier take into account the title, when provided, if no identifier was given.

----

According to [the doc](https://pandoc-numbering.readthedocs.io/en/latest/numbering.html),

> If there is no **name** and **prefix** parts, the identifier of the `span` element is constructed using the conversion of the **description** into an auto-identifier (**category**) followed by a` :` character and by the full numbering of the paragraph 

Hence, e.g.

```markdown
Problem (A) #
~ 
    This is problem A.
    
Problem (B) #
~ 
    This is problem B.
    
    
@problem:1 asks for you to…
@problem:2 is easier.
```

will produce, as expected

![image](https://user-images.githubusercontent.com/16466689/79013489-dd54e400-7b36-11ea-8d75-8dc858b0b473.png)

and 

```html
<dl>
<dt><span id="problem:1" class="pandoc-numbering-text problem"><strong>Problem 1</strong> <em>(A)</em></span></dt>
<dd>This is problem A.
</dd>
<dt><span id="problem:2" class="pandoc-numbering-text problem"><strong>Problem 2</strong> <em>(B)</em></span></dt>
<dd>This is problem B.
</dd>
</dl>
<p><a href="#problem:1" title="Problem 1 (A)"><span class="pandoc-numbering-link problem">Problem 1 (A)</span></a> asks for you to… <a href="#problem:2" title="Problem 2 (B)"><span class="pandoc-numbering-link problem">Problem 2 (B)</span></a> is easier.</p>
```
But, as the auto-identifier take into account _the number of the probem_, changing the order, swapping them, or even creating a new section, would make those references incorrect.
As one of the main purpose of this filter is to number things automatically, one does not really want to pay attention to the actual number of the environment referenced. (And I'm aware that the user can define their own identifier, but assume that doing so, when a title is already provided, can be cumbersome.)

I believe an "easy fix" would be, when the title is provided, to create the auto-identifier based on it, so that the previous example would give `<span id="problem:A" class="pandoc-numbering-text problem">` for the problem A, and similarly for B. This way, a modification in their number would not break references to them.

I don't know how easy or hard this is to implement, but I can try to have a look at it.
I know the rules to convert headers to id in [gitlab](https://gitlab.com/help/user/markdown.md#header-ids-and-links), but I don't know where similar things are implemented in pandoc. Any idea?

**Comment by aubertc** [comment]
Ok, so I had a look at it, and I believe changing the beginning of the [compute_tag](https://github.com/chdemko/pandoc-numbering/blob/develop/pandoc_numbering.py#L308) function from 

```python
    def _compute_tag(self):
        # Determine the final tag
        if self._match.group("name") is None:
            self._tag = self._category + self._number
        else:
            self._tag = self._basic_category + ":" + self._match.group("name")
```

to

```python
    def _compute_tag(self):
        # Determine the final tag
        if self._match.group("name") is None:
            if self._title is None:
                self._tag = self._category + self._number
            else:
                self._tag = self._basic_category + ":" + Numbered._identifier(stringify(Span(*self._title)))
        else:
            self._tag = self._basic_category + ":" + self._match.group("name")
```
would give what I suggest.

On a simple test, like

```markdown
Problem (A Simple Problem) #
~ 
    This is problem A.

@problem:a-simple-problem
```

it would give what I expected, namely

```html
<dl>
<dt><span id="problem:a-simple-problem" class="pandoc-numbering-text problem"><strong>Problem 1</strong> <em>(A Simple Problem)</em></span></dt>
<dd>This is problem A.
</dd>
</dl>
<p><a href="#problem:a-simple-problem" title="Problem 1 (A Simple Problem)"><span class="pandoc-numbering-link problem">Problem 1 (A Simple Problem)</span></a></p>
```

However, it conflicts with some of your tests. I can go ahead and submit a pull request if you like the idea, or simply keep it for myself if you prefer `;-)`

**Comment by aubertc** [comment]
Oh, and so, to answer my own question:

> I know the rules to convert headers to id in [gitlab](https://gitlab.com/help/user/markdown.md#header-ids-and-links), but I don't know where similar things are implemented in pandoc. Any idea?

I realized you implemented it as a static [_identifier](https://github.com/chdemko/pandoc-numbering/blob/develop/pandoc_numbering.py#L168) function, that I'm using.
However, this won't work if two titles are identical (which would be weird in any case).

Isn't there something more "standard"? I'd be surprised if that function is not already implemented in pandoc. Isn't [uniqueIdent](https://github.com/jgm/pandoc/blob/a5fa55969f1b4afc0ca3e38be50b69c65d43a460/src/Text/Pandoc/Shared.hs#L625) precisely doing that?

**Comment by chdemko** [comment]
This is a good idea.
But we have to find a way

- to maintain unique identifiers
- to keep old behavior (for compatibility)

-

**Comment by chdemko** [comment]
I propose to convert

~~~markdown
Problem (A) #
~ 
    This is problem A.
    
Problem (B) #
~ 
    This is problem B.
    
    
@problem:1 asks for you to…
@problem:2 is easier.
~~~

to

~~~markdown
[ ]{#problem:1}[ ]{#problem:a}[**Problem 1** *(A)*]{.pandoc-numbering-text .problem}
~ 
    This is problem A.
    
[ ]{#problem:2}[ ]{#problem:b}[**Problem 2** *(B)*]{.pandoc-numbering-text .problem}
~ 
    This is problem B.
    

[[Problem 1 (A)]{.pandoc-numbering-link .problem}](#problem:a "Problem 1 (A)") asks for you to... 
[[Problem 2 (B)]{.pandoc-numbering-link .problem}](#problem:b "Problem 2 (B)") is
easier.
~~~

Thus, two identifiers are created for each problem, the `#problem:1` current one and the `#problem:a` as you suggest. In case of non unique identifiers, pandoc will create a link to the first one.

**Comment by aubertc** [comment]
> But we have to find a way
> 
>     * to maintain unique identifiers

If that [uniqueIdent](https://github.com/jgm/pandoc/blob/a5fa55969f1b4afc0ca3e38be50b69c65d43a460/src/Text/Pandoc/Shared.hs#L625) function I was referring to was added to [panflute/tools.py](https://github.com/sergiocorreia/panflute/blob/master/panflute/tools.py), we would have an easy way… Any idea how hard it would be to port it?

I believe pandoc-numbering's ids will _never_ conflict with "pure" pandoc's ids, since pandoc does not generate ids _without_ colons (`:`). So, as pandoc-numbering's ids contains `:`, we'll never run into that probem.
Which means that we "only" need to worry about conflict between pandoc-numbering's ids.

>     * to keep old behavior (for compatibility)

Can't we just introduce a toggle / flag in the `pandoc-numbering:` part of the `YAML` header, and set it by default to `false`?

I'm not sure I understand why you want to have this `[ ]{#problem:1}[ ]{#problem:a}`: having empty `span`s in the htm output seems "kind of" bad to me, and will make it somewhat hard to target those elements with styling (but that's probably not too bad).

**Comment by chdemko** [comment]
> If that [uniqueIdent](https://github.com/jgm/pandoc/blob/a5fa55969f1b4afc0ca3e38be50b69c65d43a460/src/Text/Pandoc/Shared.hs#L625) function I was referring to was added to [panflute/tools.py](https://github.com/sergiocorreia/panflute/blob/master/panflute/tools.py), we would have an easy way… Any idea how hard it would be to port it?
> 
> I believe pandoc-numbering's ids will _never_ conflict with "pure" pandoc's ids, since pandoc does not generate ids _without_ colons (`:`). So, as pandoc-numbering's ids contains `:`, we'll never run into that probem.
> Which means that we "only" need to worry about conflict between pandoc-numbering's ids.
> 
We can use the `Numbered._identifier` function
> 
> Can't we just introduce a toggle / flag in the `pandoc-numbering:` part of the `YAML` header, and set it by default to `false`?
> 
> I'm not sure I understand why you want to have this `[ ]{#problem:1}[ ]{#problem:a}`: having empty `span`s in the htm output seems "kind of" bad to me, and will make it somewhat hard to target those elements with styling (but that's probably not too bad).
Using empty spans allows to refer them with a link. We can keep another span for the content and styling it using the classes.

**Comment by aubertc** [comment]
> We can use the `Numbered._identifier` function

I don't think we can: if two titles are identical, then we have a collision, and we would not have unique identifiers.
Or did I missed something?

> Using empty spans allows to refer them with a link. We can keep another span for the content and styling it using the classes.

Sure, but if you wanted to use the [target identifier](https://developer.mozilla.org/en-US/docs/Web/CSS/:target) from CSS, i.e. have something like ([with the `:` escaped](https://stackoverflow.com/a/656463) to `\3A`) `.problem\3A a:target`{.css}, then you'll target an empty span instead of the title.
That's a corner case, but would make it weird nevertheless.

**Comment by chdemko** [comment]
> Can't we just introduce a toggle / flag in the `pandoc-numbering:` part of the `YAML` header, and set it by default to `false`?

Okay, this is the right solution. The uniqueness of identifiers will be easy to maintain using the pandoc strategy. I'm going to propose a PR based upon your proposition.

**Comment by aubertc** [comment]
Actually, you may want this toggle to be three-valued:

# `default`

current behavior (i.e., that gives  `id="problem:1"` if the problem is numbered `1`, and ignore the title.)

#  `title`

The behavior I was suggesting (i.e., that gives `id="problem:a-simple-problem"` if the title is `A Simple Problem`, and  `id="problem:1"` if the problem has no title and is numbered `1`),

# `mixed`

The behavior you suggested, that would add two spans with ids reflecting the title and the number if the title is given, and goes back to default otherwise (i.e., `[ ]{#problem:1}[ ]{#problem:a-simple-problem}` if the problem has title "A Simple problem" and is numbered 1, `id="problem:1"` if the problem has no title and is numbered `1).



---

Of course, all of that would be overridden if the `id` is explicitly given.

**Comment by aubertc** [comment]
> I'm going to propose a PR based upon your proposition.

Oh, and thanks!
Let me know if I can help, for instance reviewing the code or writing new tests.

**Comment by chdemko** [comment]
Fix in 3.3.0.0
